src/commands/synthetics/deploy-tests-lib.ts

```diff
@@ -0,0 +1,93 @@
+import {getCommonAppBaseURL} from '../../helpers/app'
+
+import {APIHelper, EndpointError, formatBackendErrors, getApiHelper} from './api'
+import {DeployTestsCommandConfig, LocalTriggerConfig, MainReporter, ServerTest, TestConfig} from './interfaces'
+import {findUniqueLocalTestDefinition} from './local-test-definition'
+import {getTestConfigs} from './test'
+import {isLocalTriggerConfig} from './utils/internal'
+
+export const deployTests = async (reporter: MainReporter, config: DeployTestsCommandConfig): Promise<void> => {
+  const api = getApiHelper(config)
+
+  reporter.log('Deploying tests...\n\n')
+
+  const testConfigFromFile: TestConfig = {
+    tests: await getTestConfigs(config, reporter),
+  }
+
+  if (config.publicIds.length > 0) {
+    testConfigFromFile.tests = testConfigFromFile.tests.filter(
+      (test) =>
+        isLocalTriggerConfig(test) &&
+        test.local_test_definition.public_id &&
+        config.publicIds.includes(test.local_test_definition.public_id)
+    )
+  }
+
+  const publicIds = new Set(
+    testConfigFromFile.tests.flatMap((test) =>
+      isLocalTriggerConfig(test) && test.local_test_definition.public_id ? [test.local_test_definition.public_id] : []
+    )
+  )
+
+  for (const publicId of publicIds) {
+    const test = findUniqueLocalTestDefinition(testConfigFromFile, publicId)
+
+    try {
+      await deployLocalTestDefinition(api, test)
+
+      // SYNTH-17840: the edit test endpoint should return a version in the response, so we can print it in the logs and see it in version history
+      const baseUrl = getCommonAppBaseURL(config.datadogSite, config.subdomain)
+      const testLink = `${baseUrl}synthetics/details/${publicId}`
+
+      reporter.log(`New version successfully deployed for main test definition ${publicId}:\n  ⎋ ${testLink}\n\n`)
+    } catch (e) {
+      const errorMessage = formatBackendErrors(e)
+      throw new EndpointError(
+        `[${publicId}] Failed to update main test definition: ${errorMessage}\n`,
+        e.response?.status
+      )
+    }
+  }
+
+  reporter.log(`Deployed local test definitions defined in ${config.files[0]}\n`)
+}
+
+const deployLocalTestDefinition = async (api: APIHelper, test: LocalTriggerConfig): Promise<void> => {
+  // SYNTH-18434: public ID should be made required
+  const publicId = test.local_test_definition.public_id!
+
+  const existingRemoteTest = await api.getTest(publicId)
+
+  // XXX: We should detect noop updates in the backend
+  const newRemoteTest = removeUnsupportedEditTestFields({
+    ...existingRemoteTest,
+    ...test.local_test_definition,
+    config: {
+      ...existingRemoteTest.config,
+      ...test.local_test_definition.config,
+    },
+    options: {
+      ...existingRemoteTest.options,
+      ...test.local_test_definition.options,
+    },
+  })
```

Comment:
Why can't we let the backend do the merge?

Reply:
I wanted to, but the edit endpoint **requires** some properties that are not in LTDs (e.g. `options.tick_every`).

Basically that endpoint doesn't fully support partial updates 😕

Reply:
Can we update the endpoint then? 😁

Reply:
Created [a card](https://datadoghq.atlassian.net/browse/SYNTH-18528) 👌